clusterstate/clusterstate.go

```diff
@@ -0,0 +1,130 @@
+package clusterstate
+
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+
+	"github.com/giantswarm/e2etests/clusterstate/provider"
+)
+
+type Config struct {
+	Logger   micrologger.Logger
+	Provider provider.Interface
+}
+
+type ClusterState struct {
+	logger   micrologger.Logger
+	provider provider.Interface
+}
+
+func New(config Config) (*ClusterState, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+	if config.Provider == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Provider must not be empty", config)
+	}
+
+	s := &ClusterState{
+		logger:   config.Logger,
+		provider: config.Provider,
+	}
+
+	return s, nil
+}
+
+func (c *ClusterState) Test(ctx context.Context) error {
+	var err error
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", "installing e2e-app")
+
+		err = c.provider.InstallTestApp()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", "installed e2e-app")
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", "waiting for guest cluster")
+
+		err = c.provider.WaitForGuestReady()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", "guest cluster ready")
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", "rebooting master")
+
+		err = c.provider.RebootMaster()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", "rebooted master")
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", "waiting api to go down")
+
+		err = c.provider.WaitForAPIDown()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", "api is down")
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", "waiting for guest cluster")
+
+		err = c.provider.WaitForGuestReady()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", "guest cluster ready")
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", "replacing master node")
+
+		err = c.provider.ReplaceMaster()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", "master node replaced")
```

Comment:
New step is we replace the master by going to the next largest VM. Using instance type on EC2 and VM Size on Azure.

Not sure how to do this on KVM? Delete the master pod?

Reply:
Maybe some @giantswarm/sre feedback would be good here. Maybe it is just good enough to delete the host cluster pod. I am wondering though how we could maybe enforce the pod to move to another host cluster machine so we do not re-schedule in the same place.

Reply:
Yes some SRE help here would be much appreciated.

Moving the pod to another node would be ideal IMO. I'll have a look for if we can do that in K8s. That would test that the cluster state can be safely moved.

Reply:
@xh3b4sd I thought some more on how we can force the pod to change nodes.

We could cordon the node, delete the master pod and then uncordon the node. But that could be disruptive to the host cluster. So it might depend where we will run the tests.

Reply:
Uncertain. IIRC the KVM test env for CI had only one node in the first setup or something. Lets iterate here.

Reply:
OK I didn't realise that and yes won't work in that case. Yes I'd say lets go with this and iterate on it later.